File: CsvEditor.java
// CsvEditor.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class CsvEditor {
    private String filename;
    private List<String> headers;
    private List<List<String>> rows;
    private Scanner scanner;

    public CsvEditor(String filename) {
        this.filename = filename;
        headers = new ArrayList<>();
        rows = new ArrayList<>();
        scanner = new Scanner(System.in);
        if (filename != null && !filename.isEmpty() && Files.exists(Paths.get(filename)))
            load(filename);
        else
            createEmpty();
    }

    private void load(String filename) {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            if ((line = br.readLine()) != null) {
                headers = Arrays.asList(line.split(","));
                headers.replaceAll(String::trim);
            }
            while ((line = br.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    List<String> row = Arrays.asList(line.split(","));
                    row.replaceAll(String::trim);
                    rows.add(new ArrayList<>(row));
                }
            }
        } catch (IOException e) {
            System.out.println("Error loading file: " + e.getMessage());
        }
    }

    private void save(String filename) {
        if (filename == null || filename.isEmpty()) filename = this.filename;
        if (filename == null || filename.isEmpty()) {
            System.out.print("Enter filename to save: ");
            filename = scanner.nextLine().trim();
            if (!filename.endsWith(".csv")) filename += ".csv";
        }
        try (PrintWriter pw = new PrintWriter(new FileWriter(filename))) {
            pw.println(String.join(",", headers));
            for (List<String> row : rows)
                pw.println(String.join(",", row));
            System.out.println("Saved.");
        } catch (IOException e) {
            System.out.println("Error saving: " + e.getMessage());
        }
    }

    private void createEmpty() {
        System.out.println("Creating new CSV file.");
        System.out.print("Enter column headers (comma-separated): ");
        String input = scanner.nextLine().trim();
        if (input.isEmpty()) {
            headers = Arrays.asList("Column1", "Column2", "Column3");
        } else {
            headers = Arrays.asList(input.split(","));
            headers.replaceAll(String::trim);
        }
        rows = new ArrayList<>();
        save(filename != null ? filename : "new.csv");
    }

    private void display() {
        if (headers.isEmpty()) {
            System.out.println("No data loaded.");
            return;
        }
        int[] colWidths = headers.stream().mapToInt(String::length).toArray();
        for (List<String> row : rows) {
            for (int i = 0; i < row.size() && i < colWidths.length; i++) {
                if (row.get(i).length() > colWidths[i]) colWidths[i] = row.get(i).length();
            }
        }
        String sep = " | ";
        System.out.println();
        StringBuilder headerLine = new StringBuilder();
        for (int i = 0; i < headers.size(); i++) {
            headerLine.append(String.format("%-" + colWidths[i] + "s", headers.get(i)));
            if (i < headers.size() - 1) headerLine.append(sep);
        }
        System.out.println(headerLine);
        System.out.println("-".repeat(headerLine.length()));
        for (List<String> row : rows) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < headers.size(); i++) {
                String cell = i < row.size() ? row.get(i) : "";
                sb.append(String.format("%-" + colWidths[i] + "s", cell));
                if (i < headers.size() - 1) sb.append(sep);
            }
            System.out.println(sb);
        }
    }

    private void addRow() {
        if (headers.isEmpty()) {
            System.out.println("No headers defined. Please add columns first.");
            return;
        }
        List<String> row = new ArrayList<>();
        for (String h : headers) {
            System.out.print("Enter value for '" + h + "': ");
            row.add(scanner.nextLine().trim());
        }
        rows.add(row);
        System.out.println("Row added.");
    }

    private void editCell() {
        display();
        System.out.print("Enter row index (1-based): ");
        int rowIdx;
        try {
            rowIdx = Integer.parseInt(scanner.nextLine().trim()) - 1;
            if (rowIdx < 0 || rowIdx >= rows.size()) throw new NumberFormatException();
        } catch (NumberFormatException e) {
            System.out.println("Invalid row index.");
            return;
        }
        System.out.println("Columns: " + String.join(", ", headers));
        System.out.print("Enter column name: ");
        String colName = scanner.nextLine().trim();
        int colIdx = headers.indexOf(colName);
        if (colIdx == -1) {
            System.out.println("Column not found.");
            return;
        }
        System.out.print("Enter new value for '" + colName + "': ");
        String newVal = scanner.nextLine().trim();
        rows.get(rowIdx).set(colIdx, newVal);
        System.out.println("Cell updated.");
    }

    private void deleteRow() {
        display();
        System.out.print("Enter row index to delete (1-based): ");
        int rowIdx;
        try {
            rowIdx = Integer.parseInt(scanner.nextLine().trim()) - 1;
            if (rowIdx < 0 || rowIdx >= rows.size()) throw new NumberFormatException();
        } catch (NumberFormatException e) {
            System.out.println("Invalid row index.");
            return;
        }
        rows.remove(rowIdx);
        System.out.println("Row deleted.");
    }

    private void filterRows() {
        if (rows.isEmpty()) {
            System.out.println("No rows to filter.");
            return;
        }
        System.out.println("Columns: " + String.join(", ", headers));
        System.out.print("Enter column name to filter: ");
        String colName = scanner.nextLine().trim();
        int colIdx = headers.indexOf(colName);
        if (colIdx == -1) {
            System.out.println("Column not found.");
            return;
        }
        System.out.print("Enter value to match: ");
        String value = scanner.nextLine().trim();
        System.out.print("Case-sensitive? (y/n): ");
        boolean caseSensitive = scanner.nextLine().trim().equalsIgnoreCase("y");
        List<List<String>> filtered = new ArrayList<>();
        for (List<String> row : rows) {
            if (colIdx < row.size()) {
                String cell = row.get(colIdx);
                boolean match = caseSensitive ? cell.equals(value) : cell.equalsIgnoreCase(value);
                if (match) filtered.add(row);
            }
        }
        if (filtered.isEmpty()) {
            System.out.println("No matching rows.");
        } else {
            List<List<String>> oldRows = rows;
            rows = filtered;
            display();
            rows = oldRows;
        }
    }

    private void sortRows() {
        if (rows.isEmpty()) {
            System.out.println("No rows to sort.");
            return;
        }
        System.out.println("Columns: " + String.join(", ", headers));
        System.out.print("Enter column name to sort by: ");
        String colName = scanner.nextLine().trim();
        int colIdx = headers.indexOf(colName);
        if (colIdx == -1) {
            System.out.println("Column not found.");
            return;
        }
        System.out.print("Sort descending? (y/n): ");
        boolean reverse = scanner.nextLine().trim().equalsIgnoreCase("y");
        rows.sort((a, b) -> {
            String va = colIdx < a.size() ? a.get(colIdx) : "";
            String vb = colIdx < b.size() ? b.get(colIdx) : "";
            return reverse ? vb.compareTo(va) : va.compareTo(vb);
        });
        System.out.println("Rows sorted.");
    }

    private void addColumn() {
        System.out.print("Enter new column name: ");
        String colName = scanner.nextLine().trim();
        if (colName.isEmpty()) {
            System.out.println("Invalid name.");
            return;
        }
        if (headers.contains(colName)) {
            System.out.println("Column already exists.");
            return;
        }
        System.out.print("Enter default value for existing rows (leave empty for empty string): ");
        String defaultVal = scanner.nextLine().trim();
        headers.add(colName);
        for (List<String> row : rows) row.add(defaultVal);
        System.out.println("Column added.");
    }

    private void removeColumn() {
        if (headers.isEmpty()) {
            System.out.println("No columns to remove.");
            return;
        }
        System.out.println("Columns: " + String.join(", ", headers));
        System.out.print("Enter column name to remove: ");
        String colName = scanner.nextLine().trim();
        int colIdx = headers.indexOf(colName);
        if (colIdx == -1) {
            System.out.println("Column not found.");
            return;
        }
        headers.remove(colIdx);
        for (List<String> row : rows) row.remove(colIdx);
        System.out.println("Column removed.");
    }

    private void summary() {
        System.out.println("File: " + (filename != null ? filename : "(new)"));
        System.out.println("Rows: " + rows.size());
        System.out.println("Columns: " + headers.size());
        if (!headers.isEmpty()) System.out.println("Headers: " + String.join(", ", headers));
    }

    public void run() {
        System.out.println("=== CSV Editor/Viewer ===");
        if (filename != null && !filename.isEmpty())
            System.out.println("Loaded: " + filename + " (" + rows.size() + " rows, " + headers.size() + " columns)");
        while (true) {
            System.out.println("\nMenu:");
            System.out.println("1. View table");
            System.out.println("2. Add row");
            System.out.println("3. Edit cell");
            System.out.println("4. Delete row");
            System.out.println("5. Filter rows");
            System.out.println("6. Sort rows");
            System.out.println("7. Add column");
            System.out.println("8. Remove column");
            System.out.println("9. Export / Save");
            System.out.println("10. Show summary");
            System.out.println("11. Exit");
            System.out.print("Choose: ");
            String choice = scanner.nextLine().trim();
            switch (choice) {
                case "1": display(); break;
                case "2": addRow(); break;
                case "3": editCell(); break;
                case "4": deleteRow(); break;
                case "5": filterRows(); break;
                case "6": sortRows(); break;
                case "7": addColumn(); break;
                case "8": removeColumn(); break;
                case "9": save(null); break;
                case "10": summary(); break;
                case "11":
                    System.out.print("Save before exit? (y/n): ");
                    if (scanner.nextLine().trim().equalsIgnoreCase("y")) save(null);
                    System.out.println("Goodbye!");
                    return;
                default: System.out.println("Invalid choice.");
            }
        }
    }

    public static void main(String[] args) {
        String filename = args.length > 0 ? args[0] : null;
        CsvEditor editor = new CsvEditor(filename);
        editor.run();
    }
}

File: csv_editor.js
// csv_editor.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

class CSVEditor {
    constructor(filename) {
        this.filename = filename;
        this.headers = [];
        this.rows = [];
        if (filename && fs.existsSync(filename)) {
            this.load(filename);
        } else if (filename) {
            this.createEmpty();
        } else {
            this.createEmpty();
        }
    }

    load(filename) {
        const content = fs.readFileSync(filename, 'utf8');
        const lines = content.split('\n').filter(line => line.trim() !== '');
        if (lines.length > 0) {
            this.headers = lines[0].split(',').map(h => h.trim());
            this.rows = lines.slice(1).map(line => line.split(',').map(c => c.trim()));
        } else {
            this.headers = [];
            this.rows = [];
        }
    }

    save(filename) {
        if (!filename) filename = this.filename;
        if (!filename) {
            console.log('No filename provided.');
            return;
        }
        const lines = [];
        lines.push(this.headers.join(','));
        this.rows.forEach(row => lines.push(row.join(',')));
        fs.writeFileSync(filename, lines.join('\n'), 'utf8');
        console.log('Saved.');
    }

    createEmpty() {
        console.log('Creating new CSV file.');
        this.askQuestion('Enter column headers (comma-separated): ', (answer) => {
            const parts = answer.split(',').map(h => h.trim()).filter(h => h);
            this.headers = parts.length ? parts : ['Column1', 'Column2', 'Column3'];
            this.rows = [];
            this.save(this.filename || 'new.csv');
            this.run();
        });
    }

    askQuestion(question, callback) {
        rl.question(question, callback);
    }

    display() {
        if (this.headers.length === 0) {
            console.log('No data loaded.');
            return;
        }
        const colWidths = this.headers.map(h => h.length);
        this.rows.forEach(row => {
            row.forEach((cell, i) => {
                if (i < colWidths.length && cell.length > colWidths[i]) {
                    colWidths[i] = cell.length;
                }
            });
        });
        const sep = ' | ';
        console.log();
        console.log(this.headers.map((h, i) => h.padEnd(colWidths[i])).join(sep));
        console.log(colWidths.map(w => '-'.repeat(w)).join('-+-'));
        this.rows.forEach(row => {
            const cells = row.map((c, i) => (i < colWidths.length ? c : '').padEnd(colWidths[i] || 0));
            console.log(cells.join(sep));
        });
    }

    addRow() {
        if (this.headers.length === 0) {
            console.log('No headers defined. Please add columns first.');
            return;
        }
        const row = [];
        const askNext = (index) => {
            if (index >= this.headers.length) {
                this.rows.push(row);
                console.log('Row added.');
                this.run();
                return;
            }
            this.askQuestion(`Enter value for '${this.headers[index]}': `, (val) => {
                row.push(val.trim());
                askNext(index + 1);
            });
        };
        askNext(0);
    }

    editCell() {
        this.display();
        this.askQuestion('Enter row index (1-based): ', (rowIdxStr) => {
            const rowIdx = parseInt(rowIdxStr) - 1;
            if (isNaN(rowIdx) || rowIdx < 0 || rowIdx >= this.rows.length) {
                console.log('Invalid row index.');
                this.run();
                return;
            }
            console.log('Columns:', this.headers.join(', '));
            this.askQuestion('Enter column name: ', (colName) => {
                const colIdx = this.headers.indexOf(colName.trim());
                if (colIdx === -1) {
                    console.log('Column not found.');
                    this.run();
                    return;
                }
                this.askQuestion(`Enter new value for '${colName}': `, (newVal) => {
                    this.rows[rowIdx][colIdx] = newVal.trim();
                    console.log('Cell updated.');
                    this.run();
                });
            });
        });
    }

    deleteRow() {
        this.display();
        this.askQuestion('Enter row index to delete (1-based): ', (idxStr) => {
            const idx = parseInt(idxStr) - 1;
            if (isNaN(idx) || idx < 0 || idx >= this.rows.length) {
                console.log('Invalid row index.');
                this.run();
                return;
            }
            this.rows.splice(idx, 1);
            console.log('Row deleted.');
            this.run();
        });
    }

    filterRows() {
        if (this.rows.length === 0) {
            console.log('No rows to filter.');
            this.run();
            return;
        }
        console.log('Columns:', this.headers.join(', '));
        this.askQuestion('Enter column name to filter: ', (colName) => {
            const colIdx = this.headers.indexOf(colName.trim());
            if (colIdx === -1) {
                console.log('Column not found.');
                this.run();
                return;
            }
            this.askQuestion('Enter value to match: ', (value) => {
                this.askQuestion('Case-sensitive? (y/n): ', (caseSens) => {
                    const sensitive = caseSens.trim().toLowerCase() === 'y';
                    const filtered = this.rows.filter(row => {
                        if (colIdx >= row.length) return false;
                        const cell = row[colIdx];
                        return sensitive ? cell === value : cell.toLowerCase() === value.toLowerCase();
                    });
                    if (filtered.length === 0) {
                        console.log('No matching rows.');
                    } else {
                        const oldRows = this.rows;
                        this.rows = filtered;
                        this.display();
                        this.rows = oldRows;
                    }
                    this.run();
                });
            });
        });
    }

    sortRows() {
        if (this.rows.length === 0) {
            console.log('No rows to sort.');
            this.run();
            return;
        }
        console.log('Columns:', this.headers.join(', '));
        this.askQuestion('Enter column name to sort by: ', (colName) => {
            const colIdx = this.headers.indexOf(colName.trim());
            if (colIdx === -1) {
                console.log('Column not found.');
                this.run();
                return;
            }
            this.askQuestion('Sort descending? (y/n): ', (reverseStr) => {
                const reverse = reverseStr.trim().toLowerCase() === 'y';
                this.rows.sort((a, b) => {
                    const valA = colIdx < a.length ? a[colIdx] : '';
                    const valB = colIdx < b.length ? b[colIdx] : '';
                    if (reverse) return valB.localeCompare(valA);
                    return valA.localeCompare(valB);
                });
                console.log('Rows sorted.');
                this.run();
            });
        });
    }

    addColumn() {
        this.askQuestion('Enter new column name: ', (colName) => {
            colName = colName.trim();
            if (!colName) {
                console.log('Invalid name.');
                this.run();
                return;
            }
            if (this.headers.includes(colName)) {
                console.log('Column already exists.');
                this.run();
                return;
            }
            this.askQuestion('Enter default value for existing rows (leave empty for empty string): ', (defaultVal) => {
                this.headers.push(colName);
                this.rows.forEach(row => row.push(defaultVal.trim()));
                console.log('Column added.');
                this.run();
            });
        });
    }

    removeColumn() {
        if (this.headers.length === 0) {
            console.log('No columns to remove.');
            this.run();
            return;
        }
        console.log('Columns:', this.headers.join(', '));
        this.askQuestion('Enter column name to remove: ', (colName) => {
            const colIdx = this.headers.indexOf(colName.trim());
            if (colIdx === -1) {
                console.log('Column not found.');
                this.run();
                return;
            }
            this.headers.splice(colIdx, 1);
            this.rows.forEach(row => row.splice(colIdx, 1));
            console.log('Column removed.');
            this.run();
        });
    }

    summary() {
        console.log(`File: ${this.filename || '(new)'}`);
        console.log(`Rows: ${this.rows.length}`);
        console.log(`Columns: ${this.headers.length}`);
        if (this.headers.length) console.log('Headers:', this.headers.join(', '));
        this.run();
    }

    run() {
        console.log('\n=== CSV Editor/Viewer ===');
        if (this.filename) {
            console.log(`Loaded: ${this.filename} (${this.rows.length} rows, ${this.headers.length} columns)`);
        } else {
            console.log('No file loaded.');
        }
        console.log('\nMenu:');
        console.log('1. View table');
        console.log('2. Add row');
        console.log('3. Edit cell');
        console.log('4. Delete row');
        console.log('5. Filter rows');
        console.log('6. Sort rows');
        console.log('7. Add column');
        console.log('8. Remove column');
        console.log('9. Export / Save');
        console.log('10. Show summary');
        console.log('11. Exit');
        this.askQuestion('Choose: ', (choice) => {
            switch (choice.trim()) {
                case '1': this.display(); this.run(); break;
                case '2': this.addRow(); break;
                case '3': this.editCell(); break;
                case '4': this.deleteRow(); break;
                case '5': this.filterRows(); break;
                case '6': this.sortRows(); break;
                case '7': this.addColumn(); break;
                case '8': this.removeColumn(); break;
                case '9': this.save(); this.run(); break;
                case '10': this.summary(); break;
                case '11':
                    this.askQuestion('Save before exit? (y/n): ', (save) => {
                        if (save.trim().toLowerCase() === 'y') this.save();
                        console.log('Goodbye!');
                        rl.close();
                    });
                    break;
                default: console.log('Invalid choice.'); this.run(); break;
            }
        });
    }
}

function main() {
    const filename = process.argv[2];
    const editor = new CSVEditor(filename);
    // If createEmpty was called, it calls run internally; otherwise we start run.
    // But the constructor calls createEmpty which calls run, so we don't call again.
}

main();
